### PR TITLE
Feat/custom type node ref

### PR DESCRIPTION
**Pull request by vuhainguyen**
CDF-17207

### CODE REVIEW

File: src/test/scala/cognite/spark/v1/DataModelInstancesRelationTest.scala [hunk 1]
```diff
@@ -1115,4 +1142,93 @@ class DataModelInstancesRelationTest
...
+    ex.getMessage shouldBe "Direct relation identifier should be named_struct(\"spaceExternalId\", yourSpace, \"externalId\", yourExternalId) or " +
+      "named_struct(\"externalId\", yourExternalId) but got notExternalId as the schema."
+
+    val ex1 = sparkIntercept {
+      insertRows(
+        primitiveExtId2,
+        spark
+          .sql(s"""
+                  |select named_struct("spaceExternalId", "vu", "notExternalId", "hai") as prop_direct_relation,
+                  |'hello_my_name_is_emel' as externalId""".stripMargin)
```
Comment (hakontro) [inline src/test/scala/cognite/spark/v1/DataModelInstancesRelationTest.scala:1213]:
```suggestion
                  |'hello_my_name_is_håkon' as externalId""".stripMargin)
```

Comment (vvemel) [reply src/test/scala/cognite/spark/v1/DataModelInstancesRelationTest.scala]: haha :D

File: src/test/scala/cognite/spark/v1/DataModelInstancesRelationTest.scala [hunk 2]
```diff
@@ -1080,9 +1080,36 @@ class DataModelInstancesRelationTest
         props.get("prop_direct_relation").map(_.value) shouldBe None
       }
     )
+
+    tryTestAndCleanUp(
+      Seq(randomId), {
+        retryWhile[Boolean](
```
Comment (TharinduDG) [inline src/test/scala/cognite/spark/v1/DataModelInstancesRelationTest.scala:1086]: I don't understand the logic behind this retry function.
If there's an issue, it should fail fast and we'll know there's an issue.
I think this retry just hide errors.

**Review by TharinduDG — APPROVED** [review]
LGTM except the `retryWhile` 😉

File: src/test/scala/cognite/spark/v1/DataModelInstancesRelationTest.scala (hunk 2 above)
Comment (vuhainguyen) [reply src/test/scala/cognite/spark/v1/DataModelInstancesRelationTest.scala:1086]: it helps to retry (max 20  times with backoff delay) and is used in many places of Spark DS
https://github.com/search?l=Scala&q=org%3Acognitedata+retryWhile&type=Code
When the api is not stable enough I think it's just fine to retry it anyway, we can reduce the number of retries though

**Pull request merged**